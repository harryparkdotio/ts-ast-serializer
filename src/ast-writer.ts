import * as ts from 'typescript';

export const astWriter = (ast: any): string => {
  const sf = ts.createSourceFile('snapshot.ts', '', ts.ScriptTarget.Latest);

  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
    removeComments: false,
  });

  return printer.printNode(ts.EmitHint.Unspecified, ast, sf);
};
