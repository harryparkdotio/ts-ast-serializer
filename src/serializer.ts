import { SyntaxKind } from 'typescript';

import { astWriter } from './ast-writer';

export const serializer: jest.SnapshotSerializerPlugin = {
  test: (value: any) => {
    return (
      value && value.kind && Object.values(SyntaxKind).includes(value.kind)
    );
  },
  print: (value: any) => {
    return astWriter(value);
  },
};
