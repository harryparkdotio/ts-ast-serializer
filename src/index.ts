import { serializer } from './serializer';

module.exports = serializer;
