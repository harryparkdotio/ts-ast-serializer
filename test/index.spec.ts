import * as ts from 'typescript';

import { serializer } from '../src/serializer';

expect.addSnapshotSerializer(serializer);

it('serializes type alias', () => {
  const node = ts.createTypeAliasDeclaration(
    undefined,
    undefined,
    'abc',
    undefined,
    ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
  );

  expect(node).toMatchSnapshot();
});

it('serializes enum', () => {
  const node = ts.createEnumDeclaration(
    undefined,
    [ts.createToken(ts.SyntaxKind.ExportKeyword)],
    ts.createIdentifier('WhatAnEnum'),
    ['RED', 'GREEN', 'BLUE'].map(value =>
      ts.createEnumMember(value, ts.createStringLiteral(value))
    )
  );

  expect(node).toMatchSnapshot();
});

it('serializes string literal + comment', () => {
  const node = ts.addSyntheticLeadingComment(
    ts.createStringLiteral('abcd'),
    ts.SyntaxKind.SingleLineCommentTrivia,
    'this is a comment',
    true
  );

  expect(node).toMatchSnapshot();
});
