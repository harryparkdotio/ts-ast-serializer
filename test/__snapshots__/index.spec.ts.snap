// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializes enum 1`] = `
export enum WhatAnEnum {
    RED = "RED",
    GREEN = "GREEN",
    BLUE = "BLUE"
}
`;

exports[`serializes string literal + comment 1`] = `
//this is a comment
"abcd"
`;

exports[`serializes type alias 1`] = `type abc = any;`;
